### PR TITLE
Minor: Fix lifecycle of AdminClient and KafkaTopicClientImpl instances.

### PR DESCRIPTION
**Pull request by big-andy-coates**
The test logs where being spammed with huge volumes of warnings from Admin client instances that had not been closed, likely increasing build time. Looking at the code, there seemed to be a lot of confusion about who owned what resource and who was responsible for closing it.

A solid rule is that a class should not `close` a resource it did not create. Switched AdminClient and KafkaTopicClient instances over to this pattern, fixing the test issues and cleaning up the code.

### CODE REVIEW

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java [hunk 1]
```diff
@@ -102,29 +105,57 @@
...
+        false,
+        schemaRegistryClient,
+        new MetaStoreImpl()
+    );
+  }
+
+  KsqlEngine(
+      final KsqlConfig ksqlConfig,
+      final KafkaTopicClient topicClient,
+      final boolean topicClientOwned,
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: This is just so we know if to close the topicClient? Seems off to me.

Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: I've refactored to remove the need for it. (Though it still requires a slightly hacky test-only derived class gain access to the protected constructor. This is mainly as I don't want to expose public constructors that take in a topicClient, as doing so then means the engine might not _own_ the topic client again).

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java [hunk 2]
```diff
@@ -414,6 +400,22 @@ public static KsqlRestApplication buildApplication(
...
+
+      return client.describeCluster().clusterId().get();
+
+    } catch (final UnsupportedVersionException e) {
+      throw new KsqlException(
+          "The kafka brokers are incompatible with. "
+          + "KSQL requires broker versions >= 0.10.1.x"
+      );
+    } catch (final Exception e) {
+      throw new KsqlException("Failed to get Kafka cluster information", e);
```
Comment (rodesai) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java]: Why have a catch-all here?

Comment (rodesai) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java]: Whoops I thought I deleted this comment. Having a human-friendly error message here makes sense.

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java [hunk 3]
```diff
@@ -102,29 +104,40 @@
...
+  ) {
+    this(
+        ksqlConfig,
+        new KafkaTopicClientImpl(ksqlConfig.getKsqlAdminClientConfigProps()),
+        schemaRegistryClient,
+        new MetaStoreImpl()
+    );
+  }
+
+  protected KsqlEngine(
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: could this be `package-private`?

**Pull request merged**